packages/chevrotain/src/parse/grammar/lookahead_public.ts

```diff
@@ -0,0 +1,160 @@
+import {
+  ILookaheadStrategy,
+  ILookaheadValidationError,
+  IOrAlt,
+  LookaheadSequence,
+  LookaheadProductionType,
+  Rule,
+  TokenType,
+  OptionalProductionType
+} from "@chevrotain/types"
+import flatMap from "lodash/flatMap"
+import isEmpty from "lodash/isEmpty"
+import { defaultGrammarValidatorErrorProvider } from "../errors_public"
+import {
+  validateAmbiguousAlternationAlternatives,
+  validateEmptyOrAlternative,
+  validateNoLeftRecursion,
+  validateSomeNonEmptyLookaheadPath
+} from "./checks"
+import {
+  buildAlternativesLookAheadFunc,
+  buildLookaheadFuncForOptionalProd,
+  buildLookaheadFuncForOr,
+  buildSingleAlternativeLookaheadFunction,
+  getLookaheadPathsForOptionalProd,
+  getLookaheadPathsForOr,
+  getProdType,
+  PROD_TYPE
+} from "./lookahead"
+import { IParserDefinitionError } from "./types"
+
+export function getLookaheadPaths(options: {
+  occurrence: number
+  rule: Rule
+  prodType: LookaheadProductionType
+  maxLookahead: number
+}): LookaheadSequence[] {
+  const { occurrence, rule, prodType, maxLookahead } = options
+  const type = getProdType(prodType)
+  if (type === PROD_TYPE.ALTERNATION) {
+    return getLookaheadPathsForOr(occurrence, rule, maxLookahead)
+  } else {
+    return getLookaheadPathsForOptionalProd(
+      occurrence,
+      rule,
+      type,
+      maxLookahead
+    )
+  }
+}
+
+export class LLkLookaheadStrategy implements ILookaheadStrategy {
+  validate(options: {
+    rules: Rule[]
+    maxLookahead: number
```

Comment:
how come there is "only" a **single** `maxlookahead`?
Afaik specific productions can customize the maxLookahead
so how would this work? or is this a per-existing bug?
Or alternatively am I missing something?

Reply:
Yeah, we only pass the global `maxLookahead` into the validate function. The function itself will then drill into each rule and determine the correct `maxLookahead` at each level (grammar/rule/production). This is also the way it worked previously, but it's not a bug. It works as intended.

Reply:
okay, great